File: src/main/java/com/anod/appwatcher/search/ResultsAdapter.java
package com.anod.appwatcher.search;

import android.content.Context;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.anod.appwatcher.R;
import com.anod.appwatcher.model.WatchAppList;
import com.anod.appwatcher.utils.InstalledAppsProvider;
import com.google.android.finsky.api.model.Document;
import com.google.android.finsky.protos.nano.Messages.Common;
import com.google.android.finsky.protos.nano.Messages.AppDetails;
import com.squareup.picasso.Picasso;

public abstract class ResultsAdapter extends RecyclerView.Adapter<ResultsAppViewHolder> {
    private final Context mContext;
    private final int mColorBgDisabled;
    private final int mColorBgNormal;
    private final WatchAppList mNewAppHandler;
    private final InstalledAppsProvider mInstalledAppsProvider;

    protected ResultsAdapter(Context context, WatchAppList newAppHandler) {
        super();
        mContext = context;
        mNewAppHandler = newAppHandler;

        mColorBgDisabled = ContextCompat.getColor(context, R.color.row_inactive);
        mColorBgNormal = ContextCompat.getColor(context, R.color.item_background);

        mInstalledAppsProvider = new InstalledAppsProvider.MemoryCache(new InstalledAppsProvider.PackageManager(context.getPackageManager()));
    }

    @Override
    public ResultsAppViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(mContext).inflate(R.layout.list_item_market_app, parent, false);
        return new ResultsAppViewHolder(view, mNewAppHandler);
    }

    public abstract Document getDocument(int position);

    @Override
    public void onBindViewHolder(ResultsAppViewHolder holder, int position) {
        Document doc = getDocument(position);

        AppDetails app = doc.getAppDetails();
        String uploadDate = app == null ? "" : app.uploadDate;
        String packageName =  app == null ? "" : app.packageName;

        holder.doc = doc;
        holder.title.setText(doc.getTitle());
        holder.details.setText(doc.getCreator());
        holder.updated.setText(uploadDate);


        if (mNewAppHandler.contains(packageName)) {
            holder.row.setBackgroundColor(mColorBgDisabled);
        } else {
            holder.row.setBackgroundColor(mColorBgNormal);
        }

        String imageUrl = doc.getIconUrl();

        Picasso.with(mContext).load(imageUrl)
                .placeholder(R.drawable.ic_blur_on_black_48dp)
                .into(holder.icon);

        boolean isInstalled = mInstalledAppsProvider.getInfo(packageName).isInstalled();
        if (isInstalled) {
            holder.price.setText(R.string.installed);
        } else {
            Common.Offer offer = doc.getOffer(Common.Offer.TYPE_1);
            if (offer == null)
            {
                holder.price.setText("");
            } else if (offer.micros == 0) {
                holder.price.setText(R.string.free);
            } else {
                holder.price.setText(offer.formattedAmount);
            }
        }
    }

    public boolean isEmpty() {
        return getItemCount() == 0;
    }
}

File: src/main/java/com/anod/appwatcher/search/ResultsAdapter.kt
package com.anod.appwatcher.search

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import com.anod.appwatcher.R
import com.anod.appwatcher.model.WatchAppList
import com.squareup.picasso.Picasso
import finsky.api.model.Document
import info.anodsplace.framework.app.ThemeCompat
import info.anodsplace.framework.content.InstalledApps

/**
 *  @author alex
 *  @date 6/3/2017
 */

abstract class ResultsAdapter(
        private val context: Context,
        private val watchAppList: WatchAppList): RecyclerView.Adapter<ResultsAppViewHolder>() {

    private val colorBgDisabled = ThemeCompat.getColor(context, R.attr.inactiveRow)
    private val colorBgNormal = ThemeCompat.getColor(context, R.attr.colorItemBackground)
    private val installedApps = InstalledApps.MemoryCache(InstalledApps.PackageManager(context.packageManager))

    val isEmpty: Boolean
        get() = this.itemCount == 0

    val isNotEmpty: Boolean
        get() = this.itemCount > 0

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ResultsAppViewHolder {
        val view = LayoutInflater.from(context).inflate(R.layout.list_item_market_app, parent, false)
        return ResultsAppViewHolder(view, watchAppList)
    }

    abstract fun document(position: Int): Document

    override fun onBindViewHolder(holder: ResultsAppViewHolder, position: Int) {
        val doc = document(position)

        val app = doc.appDetails
        val uploadDate = app.uploadDate
        val packageName = app.packageName

        holder.doc = doc
        holder.title.text = doc.title
        holder.details.text = doc.creator
        holder.updated.text = uploadDate

        if (watchAppList.contains(packageName)) {
            holder.row.setBackgroundColor(colorBgDisabled)
        } else {
            holder.row.setBackgroundColor(colorBgNormal)
        }

        Picasso.get().load(doc.iconUrl)
                .placeholder(R.drawable.ic_notifications_black_24dp)
                .into(holder.icon)

        val isInstalled = installedApps.packageInfo(packageName).isInstalled
        if (isInstalled) {
            holder.price.setText(R.string.installed)
        } else {
            val offer = doc.offer
            when {
                offer.offerType == 0 -> holder.price.text = ""
                offer.micros.toInt() == 0 -> holder.price.setText(R.string.free)
                else -> holder.price.text = offer.formattedAmount
            }
        }
    }
}
